### PR TITLE
tcp: allow conn pool cancel to request connection close

### PR DESCRIPTION
**Pull request by zuercher**
Allows cancellation of a pending connection via the TCP connection pool to
request that the pending connection be closed if it cannot be immediately used.

*Risk Level*: medium
*Testing*: unit tests
*Docs Changes*: n/a
*Release Notes*: n/a
*Fixes*: #4409

Signed-off-by: Stephan Zuercher <stephan@turbinelabs.io>

### CODE REVIEW

File: include/envoy/tcp/conn_pool.h [hunk 1]
```diff
@@ -21,8 +21,12 @@ class Cancellable {
 
   /**
    * Cancel the pending request.
+   * @param close if true the connection pool will close a pending connection as part of
+   *              cancellation (unless the number of pending requests exceeds the number of pending
+   *              connections). If false, the pending request is allowed to complete and becomes
+   *              available for a future connection request.
    */
-  virtual void cancel() PURE;
+  virtual void cancel(bool close = false) PURE;
```
[the remark this answers is not in the document]
Comment (zuercher) [reply include/envoy/tcp/conn_pool.h]: Sure. Will do.

File: include/envoy/tcp/conn_pool.h [hunk 2]
```diff
@@ -12,6 +12,18 @@ namespace Envoy {
...
 
+/**
+ * Controls the behavior of a canceled connection request.
+ */
+enum class CancelPolicy {
+  // By default, canceled connection requests allow a pending connection to complete and become
+  // available for a future connection request.
+  Default,
+  // When a connection request is canceled, closes a pending connection if there are more pending
+  // connections that pending connection requests.
```
Comment (mattklein123) [inline include/envoy/tcp/conn_pool.h]: s/that/than?

File: source/common/tcp/conn_pool.cc [hunk 3]
```diff
@@ -185,17 +196,28 @@ void ConnPoolImpl::onConnectionEvent(ActiveConn& conn, Network::ConnectionEvent
...
 
-void ConnPoolImpl::onPendingRequestCancel(PendingRequest& request) {
+void ConnPoolImpl::onPendingRequestCancel(PendingRequest& request,
+                                          ConnectionPool::CancelPolicy cancel_policy) {
   ENVOY_LOG(debug, "canceling pending request");
   request.removeFromList(pending_requests_);
   host_->cluster().stats().upstream_rq_cancelled_.inc();
+
+  // If the cancel requests closure of excess connections and there are more pending connections
+  // than requests, close the most recently created pending connection.
```
Comment (mattklein123) [inline source/common/tcp/conn_pool.cc]: For my understanding, the idea here is that if there are multiple pending requests, we only want to kill the matching one that this request created? What about if another busy connection becomes available and then the request uses that? Will we still have an excess that we need to kill? I'm wondering if this might be a little more complicated...

Comment (zuercher) [reply source/common/tcp/conn_pool.cc]: I think this feature can only completely prevent excess connections if the caller always prevents re-use of connections (which TCP proxy does). That results in ever connection request creating a new connection and this change makes it so that canceled connection requests don't result in a ready upstream connection with no matching downstream.

One could argue that there's another feature that causes requests in the ready_conns_ list to be closed if they haven't been assigned within some time period. I suppose that could cover this case reasonably well, if that timeout were set relatively low, since TCP proxy connections normally exist in ready_conns_ transiently.

Comment (zuercher) [reply source/common/tcp/conn_pool.cc]: That said, to support some odd-ball protocol where the server transmits without first receiving a client request, I think you want to end the connection ASAP.

Comment (mattklein123) [reply source/common/tcp/conn_pool.cc]: I'm still a little confused about the original bug report. For my understanding, can you summarize the sequence events in the bug report?

Comment (zuercher) [reply source/common/tcp/conn_pool.cc]: The original bug was non-SSL client connection to SSL listener causes tcp_proxy to open an upstream connection. The downstream is reset (because it's not using SSL) but the conn pool connection to the upstream is opened and left idle until the next time the conn pool is used.

Comment (mattklein123) [reply source/common/tcp/conn_pool.cc]: I see. And the user does not want the extra connections to be open and ready? Could they just set an idle timeout like we have for the HTTP connection pool? Or is there a potential issue around connections being reused in a way they should not be?

(I'm just trying to understand how the fix matches the problem).

Comment (zuercher) [reply source/common/tcp/conn_pool.cc]: I think the idle timeout only applies to active connections assigned to the tcp proxy, not idle connections in the pool (unless I've misunderstood something).

I think there's a minor issue around connection use (not really re-use since the upstream will never have received any data), but it's somewhat hypothetical since I can't think of a protocol where the server transmits first. Even then the conn pool will close the connection if it receives data on an unassigned connection.

It's certainly different behavior from the pre-conn-pool tcp proxy which would always close connections when the downstream went away.

Comment (mattklein123) [reply source/common/tcp/conn_pool.cc]: > but it's somewhat hypothetical since I can't think of a protocol where the server transmits first.

MySQL, sadly.

Comment (mattklein123) [reply source/common/tcp/conn_pool.cc]: Given that the connection reuse might be a real issue for MySQL, I'm wondering if the fix needs to be more targeted to basically bind a pending TCP connection to a pending request in certain cases, and then just kill the connection if canceled during connection? This is basically what tcp_proxy did before the conn pool change and I think we need to go back to that behavior? WDYT?

Comment (zuercher) [reply source/common/tcp/conn_pool.cc]: For TCP proxy, we get that behavior with this change. Connections will only every be busy (assigned, tcp proxy will close when it done with the conn) or pending. So we'll close pending connections on cancel because each cancellation will always see one more pending conn than request. It won't necessarily be the connection that was specifically triggered by the downstream, but I don't think there's any information exchange that would matter.

If you want to write a mysql proxy that reuses upstream connections, you do end up in the situation where a connection request kicks off a new connection and before it completes a busy connection becomes ready. And that's possible even without cancellations.

I think handling the mysql case means the cancel behavior here as well as not moving pending connections to ready in processIdleConnections when there's no pending request.

Normally, I'd just add a config flag to Cluster to enable this behavior, but I had wanted to make this automatic for the TCP proxy and I expect we'd want that for a mysql proxy as well. Any thoughts on how the conn pool could know or do we just make a flag and try to document it well enough that people know to set it?

Comment (mattklein123) [reply source/common/tcp/conn_pool.cc]: Honestly I think for right now, if this change takes us back to the behavior we had previously I think it's fine, and we can revisit an L7 MySQL proxy case later? I just want to make sure that L4 MySQL proxy will work. Is it worth adding any more comments around this discussion? I feel it's nuanced/complicated and I wouldn't want to lose this for future code readers?
